src/connector/src/sink/big_query.rs

```diff
@@ -82,23 +86,97 @@ pub struct BigQueryCommon {
     pub dataset: String,
     #[serde(rename = "bigquery.table")]
     pub table: String,
-    #[serde(rename = "bigquery.max_batch_rows", default = "default_max_batch_rows")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub max_batch_rows: usize,
-    #[serde(rename = "bigquery.retry_times", default = "default_retry_times")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub retry_times: usize,
     #[serde(default)] // default false
     #[serde_as(as = "DisplayFromStr")]
     pub auto_create: bool,
 }
 
-fn default_max_batch_rows() -> usize {
-    1024
+struct BigQueryFutureManager {
+    offset_queue: VecDeque<Option<TruncateOffset>>,
+    resp_stream: Pin<Box<dyn Stream<Item = Result<()>> + Send>>,
 }
+impl BigQueryFutureManager {
+    pub fn new(
+        max_future_num: usize,
+        resp_stream: impl Stream<Item = Result<()>> + Send + 'static,
+    ) -> Self {
+        let offset_queue = VecDeque::with_capacity(max_future_num);
+        Self {
+            offset_queue,
+            resp_stream: Box::pin(resp_stream),
+        }
+    }
+
+    pub fn add_offset(&mut self, offset: TruncateOffset, mut resp_num: usize) {
+        while resp_num > 1 {
+            self.offset_queue.push_back(None);
+            resp_num -= 1;
+        }
+        self.offset_queue.push_back(Some(offset));
+    }
 
-fn default_retry_times() -> usize {
-    5
+    pub async fn wait_next_offset(&mut self) -> Result<Option<TruncateOffset>> {
+        if let Some(Some(TruncateOffset::Barrier { .. })) = self.offset_queue.front() {
+            return Ok(self.offset_queue.pop_front().unwrap());
+        }
+        self.resp_stream
+            .next()
+            .await
+            .ok_or_else(|| SinkError::BigQuery(anyhow::anyhow!("end of stream")))??;
+        self.offset_queue.pop_front().ok_or_else(|| {
+            SinkError::BigQuery(anyhow::anyhow!(
+                "should have pending chunk offset when we receive new response"
+            ))
+        })
```

Comment:
Why we can only directly pop and return if the front of queue is `TruncateOffset::Barrier`? Is it possible to have other variants of `TruncateOffset` in the queue? And what happen if there are?

Could you please add some comments here?

Reply:
Since we don't go to the bg sink to create an async task when we receive a barrier, there's no need to wait here for the
And i will add comments.

Reply:
> Since we don't go to the bg sink to create an async task when we receive a barrier, there's no need to wait here for the

Can't quite get it. Could you elaborate with proper English?

Reply:
There 3 cases:
1. Some(barrier), since we won't be sending a barrier to bigquery, there is no need to wait for the. But we will record the barriers in the log store, so we need to truncate.
2. Some(chunk), since we send a chunk to bigquery and record it in the log store, so we need to truncate and wait for resp.
3. None, It means that we split a large chunk of RW into multiple requests for bigquery (for this chunk, in offset_queue, we set it to none , none ... chunk). So we need to wait for all the none and chunk resps and truncate in Some(chunk)